main.js

```diff
@@ -0,0 +1,251 @@
+//TODO:変数、メソッドをプライベート化するのかが中途半端になっている。意思決定をして修正する。
+
+class WeatherApp {
+  constructor(locationInput) {
+    this.locationInput = locationInput;
+    this._init(this.locationInput);
+  }
+
+  async _init(location) {
+    this._displayLoader();
+
+    try {
+      const locationDetails = await this._fetchLocationDetails(location)
+      this._checkLocationNumbers(locationDetails);
+
+    } catch (e) {
+      this._resetHtml();
+      new ErrorHandler(e);
+
+    } finally {
+      this._hideLoader();
+    }
+  }
+
+  async _fetchLocationDetails(location) {
+    const response = await fetch(`http://api.openweathermap.org/geo/1.0/direct?q=${location}&limit=10&appid=${APIKEY}`);
+    if (!response.ok) {
+      this._throwNewError(response.status);
+    }
+
+    return await response.json();
+  }
+
+  //もうちょっといいメソッド名がありそう
+  _checkLocationNumbers(locationDetails) {
+    if (!locationDetails.length) {
+      throw new ErrorNoCity('都市が見つかりませんでした。')
+    }
+
+    if (locationDetails.length === 1) {
+      const coordinates = {
+        lat: locationDetails[0].lat,
+        lon: locationDetails[0].lon,
+      }
+      this._getWeather(coordinates);
+    }
+
+    if (locationDetails.length >= 2) {
+      const allLocations = this._getAllLocation(locationDetails);
+      const uniqueLocations = this._removeDuplicateLocations(allLocations);
+
+      if (uniqueLocations.length === 1) {
+        this._getWeather(uniqueLocations);
+
+      } else {
+        this._renderLocationSuggestions(uniqueLocations);
+      }
+    }
```

Comment:
[IMO]
length === 1の重複を取り除きたいので、僕ならこうしますね

メソッドの中身をさっぱりさせることで
- 場所が特定できたら、APIを叩きに行く
- そうでない場合は複数都市の情報をとりに行けるようにする

というこのメソッドの役目がみえますね

```suggestion
    const locationInfos = this._removeDuplicateLocations(locationDetails.map(location => this._extractLocationInfo(location))) ;
       
    if (!locationInfos.length) {
      throw new ErrorNoCity('都市が見つかりませんでした。')
    }
    
    if (locationInfos.length === 1) {
      const coordinates = {
        lat: locationInfos[0].lat,
        lon: locationInfos[0].lon,
      }
      this._getWeather(coordinates);
    } else {
        this._renderLocationSuggestions(locationInfos);
      }
    }
```

Reply:
先にユニークなLocationsにすることで`locationInfos.length === 1`の分岐を一回分省けるというわけですね。
圧倒的に提案してもらった方がいいですね。

Reply:
https://github.com/KenJamesHioki/weather_app/pull/16/commits/ff452b379113c280ba2cbaec864afcc6e7f53721

[IMHO]
この2行を最初に持ってきて、`uniqueLocations.length`に対して分岐をするようにしました。
```
const allLocations = this._getAllLocation(locationDetails);
const uniqueLocations = this._removeDuplicateLocations(allLocations);
```
分岐の手順は提案してもらった方法を採用させてもらいました。